File: transcription-worker/queue_service.py
"""
RabbitMQ service for consuming transcription jobs
"""
import pika
import json
import time
from typing import Callable, Dict
from logger import logger
import config

class QueueService:
    def __init__(self):
        """Initialize RabbitMQ connection"""
        self.connection = None
        self.channel = None
    
    def connect(self):
        """Connect to RabbitMQ"""
        try:
            # Parse connection URL
            parameters = pika.URLParameters(config.RABBITMQ_URL)
            
            # Connect
            self.connection = pika.BlockingConnection(parameters)
            self.channel = self.connection.channel()
            
            # Set prefetch count (QoS)
            self.channel.basic_qos(prefetch_count=config.RABBITMQ_PREFETCH)
            
            # Ensure queue exists with same configuration as API service
            self.channel.queue_declare(
                queue=config.RABBITMQ_QUEUE,
                durable=True,
                arguments={
                    'x-message-ttl': 3600000,  # 1 hour TTL
                    'x-dead-letter-exchange': f'{config.RABBITMQ_EXCHANGE}.dlx'
                }
            )
            
            logger.info("✅ RabbitMQ connected")
            logger.info(f"   Queue: {config.RABBITMQ_QUEUE}")
            logger.info(f"   Prefetch: {config.RABBITMQ_PREFETCH}")
            
        except Exception as e:
            logger.error(f"❌ RabbitMQ connection failed: {str(e)}")
            raise
    
    def disconnect(self):
        """Disconnect from RabbitMQ"""
        try:
            if self.channel:
                self.channel.close()
            if self.connection:
                self.connection.close()
            logger.info("✅ RabbitMQ disconnected")
        except Exception as e:
            logger.error(f"❌ RabbitMQ disconnect error: {str(e)}")
    
    def consume(self, handler: Callable[[Dict], None]):
        """
        Start consuming jobs from queue
        
        Args:
            handler: Callback function to process each job
        """
        def callback(ch, method, properties, body):
            """Process incoming message"""
            try:
                # Parse job
                job = json.loads(body)
                logger.info(f"📥 Received job: {job.get('mediaId')} ({job.get('operation')})")
                
                # Get retry count from headers
                retry_count = 0
                if properties.headers:
                    retry_count = properties.headers.get('x-retry-count', 0)
                
                # Process job
                handler(job)
                
                # Acknowledge message (success)
                ch.basic_ack(delivery_tag=method.delivery_tag)
                logger.info(f"✅ Job completed: {job.get('mediaId')}")
                
            except Exception as e:
                logger.error(f"❌ Job failed: {str(e)}")
                
                # Retry logic
                if retry_count < config.MAX_RETRIES:
                    logger.warning(f"🔄 Retrying job (attempt {retry_count + 1}/{config.MAX_RETRIES})")
                    
                    # Wait before retry
                    time.sleep(config.RETRY_DELAY)
                    
                    # Requeue with incremented retry count
                    new_headers = properties.headers or {}
                    new_headers['x-retry-count'] = retry_count + 1
                    
                    ch.basic_publish(
                        exchange='',
                        routing_key=config.RABBITMQ_QUEUE,
                        body=body,
                        properties=pika.BasicProperties(
                            delivery_mode=2,  # Persistent
                            headers=new_headers
                        )
                    )
                    ch.basic_ack(delivery_tag=method.delivery_tag)
                else:
                    # Max retries reached - send to dead-letter queue
                    logger.error(f"💀 Max retries reached, sending to DLQ")
                    ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)
        
        # Start consuming
        logger.info(f"🎧 Listening for jobs on queue: {config.RABBITMQ_QUEUE}")
        self.channel.basic_consume(
            queue=config.RABBITMQ_QUEUE,
            on_message_callback=callback,
            auto_ack=False
        )
        
        try:
            self.channel.start_consuming()
        except KeyboardInterrupt:
            logger.info("🛑 Stopping consumer...")
            self.channel.stop_consuming()

File: media-worker/src/queue-service.ts
import amqp from 'amqplib';
import { config } from './config';
import { logger } from './logger';

export interface MediaJob {
  mediaId: string;
  userId: string;
  projectId: string;
  s3Key: string;
  operation: 'extract_audio' | 'transcribe' | 'generate_embeddings';
}

export class QueueService {
  private connection: any = null;
  private channel: any = null;

  /**
   * Connect to RabbitMQ
   */
  async connect(): Promise<void> {
    this.connection = await amqp.connect(config.rabbitmq.url);
    this.channel = await this.connection.createChannel();

    // Set prefetch count (how many jobs to process concurrently)
    await this.channel.prefetch(config.rabbitmq.prefetch);

    logger.info('✅ RabbitMQ connected');
  }

  /**
   * Disconnect from RabbitMQ
   */
  async disconnect(): Promise<void> {
    await this.channel?.close();
    await this.connection?.close();
    logger.info('✅ RabbitMQ disconnected');
  }

  /**
   * Start consuming jobs from queue
   * @param handler - Function to handle each job
   */
  async consume(handler: (job: MediaJob) => Promise<void>): Promise<void> {
    if (!this.channel) {
      throw new Error('RabbitMQ channel not initialized');
    }

    const queueName = config.rabbitmq.queue;

    // Ensure queue exists with same configuration as API service
    await this.channel.assertQueue(queueName, {
      durable: true,
      arguments: {
        'x-message-ttl': 3600000, // 1 hour TTL (matches API service)
        'x-dead-letter-exchange': 'syncsearch.dlx',
      },
    });

    logger.info(`🎧 Listening for jobs on queue: ${queueName}`);

    await this.channel.consume(queueName, async (msg: any) => {
      if (!msg) return;

      try {
        const job: MediaJob = JSON.parse(msg.content.toString());
        logger.info(`📥 Received job: ${job.mediaId} (${job.operation})`);

        // Process the job
        await handler(job);

        // Acknowledge successful processing
        this.channel?.ack(msg);
        logger.info(`✅ Job completed: ${job.mediaId}`);
      } catch (error) {
        logger.error(`❌ Job failed: ${error}`);

        // Check if we should retry
        const retryCount = (msg.properties.headers?.['x-retry-count'] || 0) + 1;

        if (retryCount < config.worker.maxRetries) {
          // Requeue with retry count
          logger.warn(`🔄 Retrying job (attempt ${retryCount}/${config.worker.maxRetries})`);
          
          setTimeout(() => {
            this.channel?.nack(msg, false, true); // Requeue
          }, config.worker.retryDelay);
        } else {
          // Max retries reached, send to dead-letter queue
          logger.error(`💀 Max retries reached, sending to DLQ`);
          this.channel?.nack(msg, false, false); // Don't requeue
        }
      }
    });
  }

  /**
   * Publish a job to another queue
   * @param queueName - Target queue name
   * @param job - Job payload
   */
  async publishJob(queueName: string, job: MediaJob): Promise<void> {
    if (!this.channel) {
      throw new Error('RabbitMQ channel not initialized');
    }

    await this.channel.assertQueue(queueName, { durable: true });

    const message = Buffer.from(JSON.stringify(job));
    
    this.channel.sendToQueue(queueName, message, {
      persistent: true,
      contentType: 'application/json',
      timestamp: Date.now(),
    });

    logger.info(`📤 Published job to ${queueName}: ${job.mediaId}`);
  }
}
